redhat-appstudio/service-provider-integration-operator
# Data disappearance

### What does this PR do?
This PR makes the controllers switch the `SPIAccessToken`s and `SPIAccessTokenBinding`s back to awaiting phase when the 
token disappears from the storage (or if the token metadata is impossible to fetch, which effectively is the same thing).

We need to be prepared for this because we don't control the data in Vault. Additionally when switching from secrets-based
storage to Vault, this will happen to all tokens and bindings.

Note that this PR builds on top of https://github.com/redhat-appstudio/service-provider-integration-operator/pull/91

### What issues does this PR fix or reference?
N/A

### How to test this PR?
There are integration tests describing the functionality but you can manually test it like this (assuming secrets based token storage):

1. Create a binding
1. Go through OAuth so that the binding is injected
1. Delete the opaque secret with the token data
1. The generated `SPIAccessToken` object backing the binding should switch back to being in awaiting data phase and should contain no token metadata in its status
1. The binding should be back in awaiting data phase, should no longer reference the secret with the data from the token (i.e. the syncedObjectRef in its status should be empty) and the secret specified in the spec of the binding should actually no longer exist in the cluster.
